### PR TITLE
Log when publishing Ivy or Maven metadata with known incompatible features from Gradle

### PR DESCRIPTION
**Pull request by ljacomet**
Features like constraints, platforms or declared capabilities are not
supported in the ivy.xml file format.

Features like dynamic versions, with `latest.<status>` or `1.+`, and capabilities are not
supported in the pom.xml file format.

### CODE REVIEW

**Comment by ljacomet** [comment]
I have marked this as WIP because it needs further testing and possible coverage of other cases.
However I am interested in feedback on the approach, the perceived value for user and the logging of that information at `LIFECYCLE` level.

I had a look at using `SingleMessageLogger` but the presence of the stacktrace makes it really scary IMO. And similarly for a `WARN` level logging.
And if as a user you do not care, due to your usage pattern - pom consumed by Gradle only for example - these notifications are just noise.

**Comment by ljacomet** [comment]
This is complete given its intended scope.

For `ivy-publish`, will log when any of the following is declared:
* Capabilities
* Constraints
* Platforms
* Dependencies without version

For `maven-publish`, will log when any of the following is declared:
* Capabilities
* Dependencies using `+` in the version
* Dependencies using `latest.<level>` where `level` is neither `release` nor `integration`

File: subprojects/maven/src/main/java/org/gradle/api/publish/maven/internal/publication/DefaultMavenPublication.java [hunk 1]
```diff
@@ -267,7 +283,38 @@ private void populateFromComponent() {
...
             }
+            if (!usageContext.getCapabilities().isEmpty()) {
+                if (unsupportedUsages == null) {
+                    unsupportedUsages = Sets.newHashSet();
+                }
+                for (Capability capability : usageContext.getCapabilities()) {
+                    unsupportedUsages.add(String.format("Declares capability %s:%s:%s", capability.getGroup(), capability.getName(), capability.getVersion()));
+                }
+            }
+        }
```
Comment (bigdaz) [inline subprojects/maven/src/main/java/org/gradle/api/publish/maven/internal/publication/DefaultMavenPublication.java]: I think we are missing validation for dependencies declared with attributes.

File: subprojects/ivy/src/main/java/org/gradle/api/publish/ivy/internal/publication/DefaultIvyPublication.java [hunk 2]
```diff
@@ -245,15 +257,54 @@ private void populateFromComponent() {
...
+        if (unsupportedUsages != null) {
+            TreeFormatter formatter = new TreeFormatter();
+            formatter.node(getDisplayName() + INCOMPATIBLE_FEATURE);
+            formatter.startChildren();
+            for (String usage : unsupportedUsages) {
+                formatter.node(usage);
+            }
+            formatter.endChildren();
+            LOG.lifecycle(formatter.toString());
+        }
```
Comment (bigdaz) [inline subprojects/ivy/src/main/java/org/gradle/api/publish/ivy/internal/publication/DefaultIvyPublication.java]: Most of this validation code can be extracted into a utility that is shared between Maven and Ivy publishing. This might be a simple builder that allows you to add an 'unsupportedUsage', and it constructs the message.

File: subprojects/ivy/src/main/java/org/gradle/api/publish/ivy/internal/publication/DefaultIvyPublication.java [hunk 3]
```diff
@@ -94,6 +103,8 @@ public int compare(UsageContext left, UsageContext right) {
             return left.getUsage().getName().compareTo(right.getUsage().getName());
         }
     };
+    @VisibleForTesting
+    public static final String INCOMPATIBLE_FEATURE = " uses dependency management feature(s) that will most likely make the produced Ivy file incomplete and/or not consumable by Ivy compatible solutions";
```
Comment (bigdaz) [inline subprojects/ivy/src/main/java/org/gradle/api/publish/ivy/internal/publication/DefaultIvyPublication.java]: This message is a bit wordy for my liking:

"contains dependencies that cannot be represented in a published [pom file | ivy descriptor]."
"contains dependencies that will produce a pom file that cannot be consumed by a Maven client"
"contains dependencies that will produce an Ivy descriptor that cannot be consumed by an Ivy client"

I'm not sure that there's a real use case for the last one.

File: subprojects/maven/src/main/java/org/gradle/api/publish/maven/internal/publication/DefaultMavenPublication.java (hunk 1 above)
Comment (ljacomet) [reply subprojects/maven/src/main/java/org/gradle/api/publish/maven/internal/publication/DefaultMavenPublication.java]: Added

File: subprojects/ivy/src/main/java/org/gradle/api/publish/ivy/internal/publication/DefaultIvyPublication.java (hunk 2 above)
Comment (ljacomet) [reply subprojects/ivy/src/main/java/org/gradle/api/publish/ivy/internal/publication/DefaultIvyPublication.java]: Done

File: subprojects/ivy/src/main/java/org/gradle/api/publish/ivy/internal/publication/DefaultIvyPublication.java (hunk 3 above)
Comment (ljacomet) [reply subprojects/ivy/src/main/java/org/gradle/api/publish/ivy/internal/publication/DefaultIvyPublication.java]: Reworked based on your suggestion

**Comment by ljacomet** [comment]
@bigdaz I have given this a go and addressed your remarks aside from having a user control to silence some warnings. Please have another look as this could still make it into 5.2
